From: Jen <jendermine@gmail.com>
Subject: [PATCH] Open any gallery when OnePlus Gallery is missing

Show a processing spinner until the capture is finished, and reload the
last capture thumbnail on resume since the quick one only comes from
OnePlus Gallery.
---
diff --git a/smali/co/aospa/camera/AnyGallery$1.smali b/smali/co/aospa/camera/AnyGallery$1.smali
new file mode 100644
index 00000000..f9314f46
--- /dev/null
+++ b/smali/co/aospa/camera/AnyGallery$1.smali
@@ -0,0 +1,145 @@
+.class Lco/aospa/camera/AnyGallery$1;
+.super Ljava/lang/Object;
+.source "AnyGallery.java"
+
+.implements Ljava/lang/Runnable;
+
+.annotation system Ldalvik/annotation/EnclosingMethod;
+    value = Lco/aospa/camera/AnyGallery;->open(Landroid/app/Activity;Landroid/net/Uri;Ljava/lang/String;)V
+.end annotation
+
+.annotation system Ldalvik/annotation/InnerClass;
+    accessFlags = 0x0
+    name = null
+.end annotation
+
+.field final synthetic val$activity:Landroid/app/Activity;
+
+.field final synthetic val$deadline:J
+
+.field final synthetic val$dialog:Landroid/app/AlertDialog;
+
+.field final synthetic val$handler:Landroid/os/Handler;
+
+.field final synthetic val$intent:Landroid/content/Intent;
+
+.field final synthetic val$uri:Landroid/net/Uri;
+
+.method constructor <init>(Landroid/app/AlertDialog;Landroid/app/Activity;Landroid/net/Uri;JLandroid/os/Handler;Landroid/content/Intent;)V
+    .registers 8
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "()V"
+        }
+    .end annotation
+
+    .line 46
+    iput-object p1, p0, Lco/aospa/camera/AnyGallery$1;->val$dialog:Landroid/app/AlertDialog;
+
+    iput-object p2, p0, Lco/aospa/camera/AnyGallery$1;->val$activity:Landroid/app/Activity;
+
+    iput-object p3, p0, Lco/aospa/camera/AnyGallery$1;->val$uri:Landroid/net/Uri;
+
+    iput-wide p4, p0, Lco/aospa/camera/AnyGallery$1;->val$deadline:J
+
+    iput-object p6, p0, Lco/aospa/camera/AnyGallery$1;->val$handler:Landroid/os/Handler;
+
+    iput-object p7, p0, Lco/aospa/camera/AnyGallery$1;->val$intent:Landroid/content/Intent;
+
+    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
+
+    return-void
+.end method
+
+.method public run()V
+    .registers 5
+
+    .line 49
+    iget-object v0, p0, Lco/aospa/camera/AnyGallery$1;->val$dialog:Landroid/app/AlertDialog;
+
+    invoke-virtual {v0}, Landroid/app/AlertDialog;->isShowing()Z
+
+    move-result v0
+
+    if-nez v0, :cond_9
+
+    return-void
+
+    .line 50
+    :cond_9
+    iget-object v0, p0, Lco/aospa/camera/AnyGallery$1;->val$activity:Landroid/app/Activity;
+
+    invoke-virtual {v0}, Landroid/app/Activity;->isFinishing()Z
+
+    move-result v0
+
+    if-nez v0, :cond_43
+
+    iget-object v0, p0, Lco/aospa/camera/AnyGallery$1;->val$activity:Landroid/app/Activity;
+
+    invoke-virtual {v0}, Landroid/app/Activity;->isDestroyed()Z
+
+    move-result v0
+
+    if-eqz v0, :cond_1a
+
+    goto :goto_43
+
+    .line 54
+    :cond_1a
+    iget-object v0, p0, Lco/aospa/camera/AnyGallery$1;->val$activity:Landroid/app/Activity;
+
+    iget-object v1, p0, Lco/aospa/camera/AnyGallery$1;->val$uri:Landroid/net/Uri;
+
+    invoke-static {v0, v1}, Lco/aospa/camera/AnyGallery;->-$$Nest$smisPending(Landroid/app/Activity;Landroid/net/Uri;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_36
+
+    invoke-static {}, Ljava/lang/System;->currentTimeMillis()J
+
+    move-result-wide v0
+
+    iget-wide v2, p0, Lco/aospa/camera/AnyGallery$1;->val$deadline:J
+
+    cmp-long v0, v0, v2
+
+    if-gez v0, :cond_36
+
+    .line 55
+    iget-object v0, p0, Lco/aospa/camera/AnyGallery$1;->val$handler:Landroid/os/Handler;
+
+    const-wide/16 v1, 0x96
+
+    invoke-virtual {v0, p0, v1, v2}, Landroid/os/Handler;->postDelayed(Ljava/lang/Runnable;J)Z
+
+    .line 56
+    return-void
+
+    .line 58
+    :cond_36
+    iget-object v0, p0, Lco/aospa/camera/AnyGallery$1;->val$dialog:Landroid/app/AlertDialog;
+
+    invoke-virtual {v0}, Landroid/app/AlertDialog;->dismiss()V
+
+    .line 59
+    iget-object v0, p0, Lco/aospa/camera/AnyGallery$1;->val$activity:Landroid/app/Activity;
+
+    iget-object v1, p0, Lco/aospa/camera/AnyGallery$1;->val$intent:Landroid/content/Intent;
+
+    invoke-static {v0, v1}, Lco/aospa/camera/AnyGallery;->-$$Nest$smstart(Landroid/app/Activity;Landroid/content/Intent;)V
+
+    .line 60
+    return-void
+
+    .line 51
+    :cond_43
+    :goto_43
+    iget-object v0, p0, Lco/aospa/camera/AnyGallery$1;->val$dialog:Landroid/app/AlertDialog;
+
+    invoke-virtual {v0}, Landroid/app/AlertDialog;->dismiss()V
+
+    .line 52
+    return-void
+.end method
diff --git a/smali/co/aospa/camera/AnyGallery.smali b/smali/co/aospa/camera/AnyGallery.smali
new file mode 100644
index 00000000..2d7e2b27
--- /dev/null
+++ b/smali/co/aospa/camera/AnyGallery.smali
@@ -0,0 +1,433 @@
+.class public final Lco/aospa/camera/AnyGallery;
+.super Ljava/lang/Object;
+.source "AnyGallery.java"
+
+.field private static final POLL_MS:J = 0x96L
+
+.field private static final TAG:Ljava/lang/String; = "OCAM_AnyGallery"
+
+.field private static final TIMEOUT_MS:J = 0x2710L
+
+.method static bridge synthetic -$$Nest$smisPending(Landroid/app/Activity;Landroid/net/Uri;)Z
+    .registers 2
+
+    invoke-static {p0, p1}, Lco/aospa/camera/AnyGallery;->isPending(Landroid/app/Activity;Landroid/net/Uri;)Z
+
+    move-result p0
+
+    return p0
+.end method
+
+.method static bridge synthetic -$$Nest$smstart(Landroid/app/Activity;Landroid/content/Intent;)V
+    .registers 2
+
+    invoke-static {p0, p1}, Lco/aospa/camera/AnyGallery;->start(Landroid/app/Activity;Landroid/content/Intent;)V
+
+    return-void
+.end method
+
+.method private constructor <init>()V
+    .registers 1
+
+    .line 25
+    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
+
+    return-void
+.end method
+
+.method private static dp(Landroid/app/Activity;I)I
+    .registers 2
+
+    .line 104
+    int-to-float p1, p1
+
+    invoke-virtual {p0}, Landroid/app/Activity;->getResources()Landroid/content/res/Resources;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Landroid/content/res/Resources;->getDisplayMetrics()Landroid/util/DisplayMetrics;
+
+    move-result-object p0
+
+    iget p0, p0, Landroid/util/DisplayMetrics;->density:F
+
+    mul-float/2addr p1, p0
+
+    invoke-static {p1}, Ljava/lang/Math;->round(F)I
+
+    move-result p0
+
+    return p0
+.end method
+
+.method private static isPending(Landroid/app/Activity;Landroid/net/Uri;)Z
+    .registers 7
+
+    .line 84
+    new-instance v0, Landroid/os/Bundle;
+
+    invoke-direct {v0}, Landroid/os/Bundle;-><init>()V
+
+    .line 85
+    const-string v1, "android:query-arg-match-pending"
+
+    const/4 v2, 0x1
+
+    invoke-virtual {v0, v1, v2}, Landroid/os/Bundle;->putInt(Ljava/lang/String;I)V
+
+    .line 86
+    const/4 v1, 0x0
+
+    :try_start_c
+    invoke-virtual {p0}, Landroid/app/Activity;->getContentResolver()Landroid/content/ContentResolver;
+
+    move-result-object p0
+
+    const-string v3, "is_pending"
+
+    filled-new-array {v3}, [Ljava/lang/String;
+
+    move-result-object v3
+
+    const/4 v4, 0x0
+
+    invoke-virtual {p0, p1, v3, v0, v4}, Landroid/content/ContentResolver;->query(Landroid/net/Uri;[Ljava/lang/String;Landroid/os/Bundle;Landroid/os/CancellationSignal;)Landroid/database/Cursor;
+
+    move-result-object p0
+    :try_end_1b
+    .catch Ljava/lang/Exception; {:try_start_c .. :try_end_1b} :catch_3d
+
+    .line 88
+    if-eqz p0, :cond_36
+
+    :try_start_1d
+    invoke-interface {p0}, Landroid/database/Cursor;->moveToFirst()Z
+
+    move-result v0
+
+    if-eqz v0, :cond_36
+
+    invoke-interface {p0, v1}, Landroid/database/Cursor;->getInt(I)I
+
+    move-result v0
+    :try_end_27
+    .catchall {:try_start_1d .. :try_end_27} :catchall_2a
+
+    if-eqz v0, :cond_36
+
+    goto :goto_37
+
+    .line 86
+    :catchall_2a
+    move-exception v0
+
+    if-eqz p0, :cond_35
+
+    :try_start_2d
+    invoke-interface {p0}, Landroid/database/Cursor;->close()V
+    :try_end_30
+    .catchall {:try_start_2d .. :try_end_30} :catchall_31
+
+    goto :goto_35
+
+    :catchall_31
+    move-exception p0
+
+    :try_start_32
+    invoke-virtual {v0, p0}, Ljava/lang/Throwable;->addSuppressed(Ljava/lang/Throwable;)V
+
+    :cond_35
+    :goto_35
+    throw v0
+
+    .line 88
+    :cond_36
+    move v2, v1
+
+    .line 89
+    :goto_37
+    if-eqz p0, :cond_3c
+
+    invoke-interface {p0}, Landroid/database/Cursor;->close()V
+    :try_end_3c
+    .catch Ljava/lang/Exception; {:try_start_32 .. :try_end_3c} :catch_3d
+
+    .line 88
+    :cond_3c
+    return v2
+
+    .line 89
+    :catch_3d
+    move-exception p0
+
+    .line 90
+    invoke-static {p1}, Ljava/lang/String;->valueOf(Ljava/lang/Object;)Ljava/lang/String;
+
+    move-result-object p1
+
+    new-instance v0, Ljava/lang/StringBuilder;
+
+    invoke-direct {v0}, Ljava/lang/StringBuilder;-><init>()V
+
+    const-string v2, "pending check failed for "
+
+    invoke-virtual {v0, v2}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object v0
+
+    invoke-virtual {v0, p1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p1
+
+    invoke-virtual {p1}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object p1
+
+    const-string v0, "OCAM_AnyGallery"
+
+    invoke-static {v0, p1, p0}, Landroid/util/Log;->w(Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)I
+
+    .line 91
+    return v1
+.end method
+
+.method public static open(Landroid/app/Activity;Landroid/net/Uri;Ljava/lang/String;)V
+    .registers 11
+
+    .line 28
+    if-eqz p0, :cond_52
+
+    if-nez p1, :cond_5
+
+    goto :goto_52
+
+    .line 29
+    :cond_5
+    new-instance v7, Landroid/content/Intent;
+
+    const-string v0, "com.android.camera.action.REVIEW"
+
+    invoke-direct {v7, v0}, Landroid/content/Intent;-><init>(Ljava/lang/String;)V
+
+    .line 30
+    invoke-virtual {v7, p1, p2}, Landroid/content/Intent;->setDataAndType(Landroid/net/Uri;Ljava/lang/String;)Landroid/content/Intent;
+
+    .line 31
+    const/4 p2, 0x1
+
+    invoke-virtual {v7, p2}, Landroid/content/Intent;->addFlags(I)Landroid/content/Intent;
+
+    .line 33
+    invoke-static {p0, p1}, Lco/aospa/camera/AnyGallery;->isPending(Landroid/app/Activity;Landroid/net/Uri;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_1d
+
+    .line 34
+    invoke-static {p0, v7}, Lco/aospa/camera/AnyGallery;->start(Landroid/app/Activity;Landroid/content/Intent;)V
+
+    .line 35
+    return-void
+
+    .line 38
+    :cond_1d
+    new-instance v0, Landroid/app/AlertDialog$Builder;
+
+    invoke-direct {v0, p0}, Landroid/app/AlertDialog$Builder;-><init>(Landroid/content/Context;)V
+
+    .line 39
+    invoke-static {p0}, Lco/aospa/camera/AnyGallery;->progressView(Landroid/app/Activity;)Landroid/widget/LinearLayout;
+
+    move-result-object v1
+
+    invoke-virtual {v0, v1}, Landroid/app/AlertDialog$Builder;->setView(Landroid/view/View;)Landroid/app/AlertDialog$Builder;
+
+    move-result-object v0
+
+    .line 40
+    invoke-virtual {v0, p2}, Landroid/app/AlertDialog$Builder;->setCancelable(Z)Landroid/app/AlertDialog$Builder;
+
+    move-result-object p2
+
+    .line 41
+    invoke-virtual {p2}, Landroid/app/AlertDialog$Builder;->create()Landroid/app/AlertDialog;
+
+    move-result-object v1
+
+    .line 42
+    invoke-virtual {v1}, Landroid/app/AlertDialog;->show()V
+
+    .line 44
+    new-instance v6, Landroid/os/Handler;
+
+    invoke-static {}, Landroid/os/Looper;->getMainLooper()Landroid/os/Looper;
+
+    move-result-object p2
+
+    invoke-direct {v6, p2}, Landroid/os/Handler;-><init>(Landroid/os/Looper;)V
+
+    .line 45
+    invoke-static {}, Ljava/lang/System;->currentTimeMillis()J
+
+    move-result-wide v2
+
+    const-wide/16 v4, 0x2710
+
+    add-long/2addr v4, v2
+
+    .line 46
+    new-instance v0, Lco/aospa/camera/AnyGallery$1;
+
+    move-object v2, p0
+
+    move-object v3, p1
+
+    invoke-direct/range {v0 .. v7}, Lco/aospa/camera/AnyGallery$1;-><init>(Landroid/app/AlertDialog;Landroid/app/Activity;Landroid/net/Uri;JLandroid/os/Handler;Landroid/content/Intent;)V
+
+    const-wide/16 p0, 0x96
+
+    invoke-virtual {v6, v0, p0, p1}, Landroid/os/Handler;->postDelayed(Ljava/lang/Runnable;J)Z
+
+    .line 62
+    return-void
+
+    .line 28
+    :cond_52
+    :goto_52
+    return-void
+.end method
+
+.method private static progressView(Landroid/app/Activity;)Landroid/widget/LinearLayout;
+    .registers 8
+
+    .line 65
+    const/16 v0, 0x18
+
+    invoke-static {p0, v0}, Lco/aospa/camera/AnyGallery;->dp(Landroid/app/Activity;I)I
+
+    move-result v0
+
+    .line 66
+    new-instance v1, Landroid/widget/LinearLayout;
+
+    invoke-direct {v1, p0}, Landroid/widget/LinearLayout;-><init>(Landroid/content/Context;)V
+
+    .line 67
+    const/4 v2, 0x0
+
+    invoke-virtual {v1, v2}, Landroid/widget/LinearLayout;->setOrientation(I)V
+
+    .line 68
+    const/16 v3, 0x10
+
+    invoke-virtual {v1, v3}, Landroid/widget/LinearLayout;->setGravity(I)V
+
+    .line 69
+    invoke-virtual {v1, v0, v0, v0, v0}, Landroid/widget/LinearLayout;->setPadding(IIII)V
+
+    .line 71
+    new-instance v0, Landroid/widget/ProgressBar;
+
+    invoke-direct {v0, p0}, Landroid/widget/ProgressBar;-><init>(Landroid/content/Context;)V
+
+    .line 72
+    const/4 v4, 0x1
+
+    invoke-virtual {v0, v4}, Landroid/widget/ProgressBar;->setIndeterminate(Z)V
+
+    .line 73
+    new-instance v4, Landroid/widget/LinearLayout$LayoutParams;
+
+    const/16 v5, 0x28
+
+    invoke-static {p0, v5}, Lco/aospa/camera/AnyGallery;->dp(Landroid/app/Activity;I)I
+
+    move-result v6
+
+    invoke-static {p0, v5}, Lco/aospa/camera/AnyGallery;->dp(Landroid/app/Activity;I)I
+
+    move-result v5
+
+    invoke-direct {v4, v6, v5}, Landroid/widget/LinearLayout$LayoutParams;-><init>(II)V
+
+    invoke-virtual {v1, v0, v4}, Landroid/widget/LinearLayout;->addView(Landroid/view/View;Landroid/view/ViewGroup$LayoutParams;)V
+
+    .line 75
+    new-instance v0, Landroid/widget/TextView;
+
+    invoke-direct {v0, p0}, Landroid/widget/TextView;-><init>(Landroid/content/Context;)V
+
+    .line 76
+    const-string v4, "Processing photo\u2026"
+
+    invoke-virtual {v0, v4}, Landroid/widget/TextView;->setText(Ljava/lang/CharSequence;)V
+
+    .line 77
+    const/4 v4, 0x2
+
+    const/high16 v5, 0x41800000    # 16.0f
+
+    invoke-virtual {v0, v4, v5}, Landroid/widget/TextView;->setTextSize(IF)V
+
+    .line 78
+    invoke-static {p0, v3}, Lco/aospa/camera/AnyGallery;->dp(Landroid/app/Activity;I)I
+
+    move-result p0
+
+    invoke-virtual {v0, p0, v2, v2, v2}, Landroid/widget/TextView;->setPadding(IIII)V
+
+    .line 79
+    invoke-virtual {v1, v0}, Landroid/widget/LinearLayout;->addView(Landroid/view/View;)V
+
+    .line 80
+    return-object v1
+.end method
+
+.method private static start(Landroid/app/Activity;Landroid/content/Intent;)V
+    .registers 4
+
+    .line 97
+    :try_start_0
+    invoke-virtual {p0, p1}, Landroid/app/Activity;->startActivity(Landroid/content/Intent;)V
+    :try_end_3
+    .catch Ljava/lang/Exception; {:try_start_0 .. :try_end_3} :catch_4
+
+    .line 100
+    goto :goto_21
+
+    .line 98
+    :catch_4
+    move-exception p0
+
+    .line 99
+    invoke-static {p1}, Ljava/lang/String;->valueOf(Ljava/lang/Object;)Ljava/lang/String;
+
+    move-result-object p1
+
+    new-instance v0, Ljava/lang/StringBuilder;
+
+    invoke-direct {v0}, Ljava/lang/StringBuilder;-><init>()V
+
+    const-string v1, "no gallery for "
+
+    invoke-virtual {v0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object v0
+
+    invoke-virtual {v0, p1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p1
+
+    invoke-virtual {p1}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object p1
+
+    const-string v0, "OCAM_AnyGallery"
+
+    invoke-static {v0, p1, p0}, Landroid/util/Log;->e(Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)I
+
+    .line 101
+    :goto_21
+    return-void
+.end method
diff --git a/smali/com/oplus/camera/CameraManager$f.smali b/smali/com/oplus/camera/CameraManager$f.smali
index d2a52e4a..42128b76 100644
--- a/smali/com/oplus/camera/CameraManager$f.smali
+++ b/smali/com/oplus/camera/CameraManager$f.smali
@@ -1982,6 +1982,14 @@
 
     if-nez v5, :cond_a
 
+    iget-object v6, v13, Lcom/oplus/camera/helper/GalleryHelper;->a:Landroid/app/Activity;
+
+    iget-object v5, v0, Lvj/h;->c:Landroid/net/Uri;
+
+    iget-object v7, v0, Lvj/h;->d:Ljava/lang/String;
+
+    invoke-static {v6, v5, v7}, Lco/aospa/camera/AnyGallery;->open(Landroid/app/Activity;Landroid/net/Uri;Ljava/lang/String;)V
+
     invoke-static {}, Lcom/oplus/camera/util/h1;->b()Lcom/oplus/camera/util/h1;
 
     move-result-object v5
diff --git a/smali/y4/t6.smali b/smali/y4/t6.smali
index b4c491b2..a2509175 100644
--- a/smali/y4/t6.smali
+++ b/smali/y4/t6.smali
@@ -287,7 +287,6 @@
     move-result p0
 
     .line 129
-    if-nez p0, :cond_4
 
     .line 130
     .line 131
